tests/cases/conformance/controlFlow/controlFlowIterationErrorsAsync.ts

```diff
@@ -0,0 +1,137 @@
+// @strict: true
+// @noEmit: true
+// @lib: esnext
+
+let cond: boolean;
+
+async function len(s: string) {
+    return s.length;
+}
+
+async function f1() {
+    let x: string | number | boolean;
+    x = "";
+    while (cond) {
+        x = await len(x);
+        x;
+    }
+    x;
+}
+
+async function f2() {
+    let x: string | number | boolean;
+    x = "";
+    while (cond) {
+        x;
+        x = await len(x);
+    }
+    x;
+}
+
+declare function foo(x: string): Promise<number>;
+declare function foo(x: number): Promise<string>;
+
+async function g1() {
+    let x: string | number | boolean;
+    x = "";
+    while (cond) {
+        x = await foo(x);
+        x;
+    }
+    x;
+}
+
+async function g2() {
+    let x: string | number | boolean;
+    x = "";
+    while (cond) {
+        x;
+        x = await foo(x);
+    }
+    x;
+}
+
+async function asNumber(x: string | number): Promise<number> {
+    return +x;
+}
+
+async function h1() {
+    let x: string | number | boolean;
+    x = "0";
+    while (cond) {
+        x = +x + 1;
+        x;
+    }
+}
+
+async function h2() {
+    let x: string | number | boolean;
+    x = "0";
+    while (cond) {
+        x = await asNumber(x) + 1;
+        x;
+    }
+}
+
+async function h3() {
+    let x: string | number | boolean;
+    x = "0";
+    while (cond) {
+        let y = await asNumber(x);
+        x = y + 1;
+        x;
+    }
+}
+
+async function h4() {
+    let x: string | number | boolean;
+    x = "0";
+    while (cond) {
+        x;
+        let y = await asNumber(x);
+        x = y + 1;
+        x;
+    }
+}
+
+// repro #51115
```

Comment:
Here, we have the new test cases - based on the reported issues. They are very same-ish, but I've figured out that it still might be worthwhile to commit all of them.